crates/compiler/can/src/desugar.rs

```diff
@@ -342,6 +332,15 @@ pub fn desugar_expr<'a>(
 
             arena.alloc(Loc { region, value })
         }
+        // desugar the sub_expression, but leave the TaskAwaitBang as this will
+        // be unwrapped later in desugar_value_def_suffixed
+        TaskAwaitBang(sub_expr) => {
+            let intermediate = arena.alloc(Loc::at(loc_expr.region, **sub_expr));
+            let new_sub_loc_expr = desugar_expr(arena, intermediate, src, line_info, module_path);
+            let new_sub_expr = arena.alloc(new_sub_loc_expr.value);
+
+            arena.alloc(Loc::at(loc_expr.region, TaskAwaitBang(new_sub_expr)))
+        }
```

Comment:
We leave the TaskAwaitBang in during de-sugaring, and "unwrap" it later so there shouldn't be any instances of this after unwrapping the suffixes.

Reply:
That seems fine, but I am a little curious why? What's the advantage to processing this later?

Reply:
I originally had it in with the de-sugaring but it was causing a few of issues. 

It's simpler when the AST has been de-sugared and there are aren't ParensAround and others to handle. 

It's different enough and also complex enough and so easier to reason about and test if it's not mixed in together. I found it difficult to make progress at first because the levels of recursion in de-sugaring were hard to track alongside the unwrapping suffixed things. 

We need the surrounding context from the Defs node and only want to unwrap top-level nodes (but not the top-level Defs itself), so keeping it separate makes it easier to so where the unwrapping suffixed parts starts and ends.